src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class DependeciesGroupsForEachTFMRule : IPackageRule
+    {
+        private HashSet<NuGetFramework> _compatNotExactMatches;
+        private HashSet<NuGetFramework> _noExactMatchesFromFile;
+        private HashSet<NuGetFramework> _noExactMatchesFromNuspec;
+
+        public string MessageFormat { get; }
+
+        public string CompatMatchFoundWarningMessageFormat { get; }
+
+        public DependeciesGroupsForEachTFMRule(string messageFormat, string compatMatchFoundWarningMessageFormat)
+        {
+            MessageFormat = messageFormat;
+            CompatMatchFoundWarningMessageFormat = compatMatchFoundWarningMessageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
```

Comment:
I find the program flow a bit difficult to follow. I feel like there isn't a good [separation of responsibilities](https://en.wikipedia.org/wiki/Single_responsibility_principle).

I propose that Validate's implementation could be:

```csharp
        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
        {
            var files = builder.GetFiles().
                Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1);
            var packageNuspec = ExtractTFMsFromNuspec(builder.GetNuspec());
            var (compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec) = CatagoriseTfms(files, packageNuspec);
            return GenerateWarnings(compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec);
        }
```

This way `CatagoriseTfms` and `GenerateWarnings` can be easily tested separately.